File: problems/337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from functools import lru_cache
class Solution:
    @lru_cache(None)
    def solve(self, node: Optional[TreeNode], canRob: bool) -> int:
        if node is None:
            return 0
        
        lnr = self.solve(node.left, True)
        rnr = self.solve(node.right, True)
        if canRob:
            lq = self.solve(node.left, False)
            rq = self.solve(node.right, False)
            return max(lq+rq+node.val, lnr+rnr)
        return lnr+rnr

    def rob(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        lq = self.solve(root.left, False)
        rq = self.solve(root.right, False)
        lnr = self.solve(root.left, True)
        rnr = self.solve(root.right, True)
        return max(lq+rq+root.val, lnr+rnr)

File: problems/337-house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<int, int> solve(TreeNode* node){
        if(node == NULL)
            return make_pair(0, 0);

        // taking left and not taking left
        pair<int, int> left = solve(node->left);
        // taking right and not taking right
        pair<int, int> right = solve(node->right);
        // rob this, have to skip next
        int rob = left.second + right.second;
        // not rob this, can skip next or take next
        int notRob = max(left.second, left.first) + max(right.second, right.first);
        return make_pair(rob + node->val, notRob);
    }
    int rob(TreeNode* root) {
        pair<int, int> answer = solve(root);
        return max(answer.first, answer.second);
    }
};